pkg/kubelet/dockershim/helpers.go

```diff
@@ -393,3 +396,48 @@ type dockerOpt struct {
 func (d dockerOpt) GetKV() (string, string) {
 	return d.key, d.value
 }
+
+// writeLimiter limits the total output written across one or more streams.
+type writeLimiter struct {
```

Comment:
Just to be sure I'm understanding, can you share more about why we need to create a new struct for limiting writes and can't use Go builtins such as `io.CopyN`?

Is it because we want to actually share the maximum amount of bytes to write across both stderr and stdout?

If we move forward with our own struct,  I wonder if we could update `writeLimiter` to make it a little clearer that its shared? For example, `sharedWriteLimiter` as a name?

Could you also share more about the decision to use `atomic` instead of a mutex around `limit`? I'm not an expert in go concurrency, so a little more context around the decision could be helpful for me :)

Reply:
yes, the limit is "shared" across both the stdout and stderr output stream, so CopyN doesn't really fit.

> For example, sharedWriteLimiter as a name

I'm not opposed to renaming this for clarity, though hopefully people would read the struct comment that I left that describes its use case.

> Could you also share more about the decision to use atomic instead of a mutex around limit?

Sure. I don't want to hold a mutex across I/O operations. But I do want concurrency-safe operations applied against the limit counter. Hence the use of sync/atomic. The downside of this approach is that concurrent writes can race and we *could* write more than the specified limit. However, given the softness of the requirements around the `LimitBytes` API, I think this is an acceptable trade-off that remains within spec.